### PR TITLE
Issue #411

### PR DESCRIPTION
**Pull request by inthar-raven**
Added radio buttons and the reaction. Using zero-indexing is the default and omits the column of all unisons.

### CODE REVIEW

File: src/views/AnalysisView.vue [hunk 1]
```diff
@@ -87,7 +89,8 @@ const matrix = computed(() => {
     props.scale.head(MAX_SCALE_SIZE).mergeOptions({
       centsFractionDigits: 1,
       decimalFractionDigits: 3,
-    })
+    }), 
+    1 - (matrixStartIndex.value)
```
Comment (frostburn) [inline src/views/AnalysisView.vue:93]: You shouldn't call a property something and then pass it as the inverse of that to another function.

Comment (inthar-raven) [reply src/views/AnalysisView.vue:93]: What I have should work fine (for issue #411), though. I *am* using just "matrixStartIndex.value" in other parts.

File: src/views/AnalysisView.vue [hunk 2]
```diff
@@ -99,10 +102,10 @@ const matrix = computed(() => {
       <table>
         <tr>
           <th></th>
-          <th v-for="i of Math.min(scale.size, MAX_SCALE_SIZE)" :key="i">
+          <th v-for="i of Math.min(scale.size-1+matrixStartIndex, MAX_SCALE_SIZE-1)" :key="i">
```
Comment (frostburn) [inline src/views/AnalysisView.vue:105]: `MAX_SCALE_SIZE` is a sanity limit that makes sure the browser doesn't go :boom: . You could rename it to something like `MAX_NUM_COLUMNS` or something. That `-1` should be there on a constant.

File: src/views/AnalysisView.vue (hunk 1 above)
Comment (inthar-raven) [reply src/views/AnalysisView.vue:93]: It's admittedly a really janky way to not display the zeroth column but just for the zero-indexed option. I should fix that.

File: src/analysis.ts [hunk 3]
```diff
@@ -103,12 +103,12 @@ export function utonalFundamental(frequencies: number[], maxDivisor = 23) {
 }
 
 // Interval matrix a.k.a the modes of a scale
-export function intervalMatrix(scale: Scale) {
+export function intervalMatrix(scale: Scale, startIndex: number) {
   const result = [];
-  const degrees = [...Array(scale.size + 1).keys()];
+  const degrees = [...Array(scale.size + 1 - startIndex).keys()];
```
Comment (frostburn) [inline src/analysis.ts:108]: I know this is my code, but this is no longer an array of `degrees`.
Just do an extra `for` loop below and collect stuff into  a `const row = []` array.

File: src/views/AnalysisView.vue [hunk 4]
```diff
@@ -19,6 +19,8 @@ const props = defineProps<{
 }>();
 
 const cellFormat = ref<"best" | "cents" | "decimal">("best");
+const indexing = ref<"zero" | "one">("zero");
+const matrixStartIndex = computed(() => indexing.value === "one" ? 1 : 0);
```
Comment (frostburn) [inline src/views/AnalysisView.vue:23]: To be specific: Swap here, not down there.

Comment (frostburn) [reply src/views/AnalysisView.vue:23]: Computed properties should be collected together after mutable state.

**Comment by frostburn** [comment]
The choice should be persisted
```
watch(indexing, (newValue) =>
  window.localStorage.setItem("intervalMatrixIndexing", newValue)
);
```
and the corresponding `.getItem(...) ?? "zero"` when initializing the ref.

File: src/views/AnalysisView.vue [hunk 5]
```diff
@@ -19,6 +19,8 @@ const props = defineProps<{
 }>();
 
 const cellFormat = ref<"best" | "cents" | "decimal">("best");
+const intervalMatrixIndexing = ref<"zero" | "one">("zero");
```
Comment (frostburn) [inline src/views/AnalysisView.vue:22]: If you want to persist the value you have to use a component prop and drill it back to the app using emits.
Now this line just creates a new local property.

File: src/App.vue [hunk 6]
```diff
@@ -288,6 +289,7 @@ watch(
     pingPongFeedback,
     pingPongSeparation,
     pingPongGain,
+    intervalMatrixIndexing
```
Comment (frostburn) [inline src/App.vue:292]: You specifically don't want to share your personal preferences with the world.

**Comment by frostburn** [comment]
The app doesn't load at all for me now.

File: src/analysis.ts [hunk 7]
```diff
@@ -103,12 +103,12 @@ export function utonalFundamental(frequencies: number[], maxDivisor = 23) {
 }
 
 // Interval matrix a.k.a the modes of a scale
-export function intervalMatrix(scale: Scale) {
+export function intervalMatrix(scale: Scale, startIndex: number) {
```
Comment (frostburn) [inline src/analysis.ts:106]: This is now an unrelated change that doesn't belong in the scope of this PR.

**Comment by frostburn** [comment]
Now it works on my end. What's the intended behaviour?
I'm seeing that toggling the radio button simply adds one to the header row. If that's all that this is supposed to do, then clean the code to only do that. No need for fancy computes or anything.

File: src/App.vue [hunk 8]
```diff
@@ -120,6 +120,7 @@ const centsFractionDigits = ref(3);
 const decimalFractionDigits = ref(5);
 const showVirtualQwerty = ref(false);
 const midiOctaveOffset = ref(-1);
+const intervalMatrixIndexing = ref<"zero" | "one">("zero");
```
Comment (frostburn) [inline src/App.vue:123]: This could be replaced with a number called `intervalMatrixIndexOffset` that gets added to the headers if that's the intended effect.

Comment (inthar-raven) [reply src/App.vue:123]: In my previous iterations of attempting that, it seems the radio buttons don't like me to make its associated type a `number` instead of a `radio`:

Internal server error: Unnecessary value binding used alongside v-model. It will interfere with v-model's behavior.
  Plugin: vite:vue
  File: /home/inthar/Downloads/scale-workshop-main/src/views/AnalysisView.vue
  160|            <input
  161|              type="number"
  162|              id="indexing-zero"
     |                               ^
  163|              value="0"
  164|              v-model="intervalMatrixIndexing"
      at Object.createCompilerError (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:19:19)
      at createDOMCompilerError (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-dom/dist/compiler-dom.cjs.js:2483:25)
      at checkDuplicatedValue (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-dom/dist/compiler-dom.cjs.js:2547:29)
      at transformModel (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-dom/dist/compiler-dom.cjs.js:2579:29)
      at buildProps (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:4777:48)
      at Array.postTransformElement (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:4374:38)
      at traverseNode (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:2225:19)
      at traverseChildren (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:2167:9)
      at traverseNode (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:2218:13)
      at traverseChildren (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:2167:9)

Comment (inthar-raven) [reply src/App.vue:123]: In my previous iterations of attempting that, it seems the radio buttons don't want me to make their associated type a `number` instead of a `radio`:

Internal server error: Unnecessary value binding used alongside v-model. It will interfere with v-model's behavior.
  Plugin: vite:vue
  File: /home/inthar/Downloads/scale-workshop-main/src/views/AnalysisView.vue
  160|            <input
  161|              type="number"
  162|              id="indexing-zero"
     |                               ^
  163|              value="0"
  164|              v-model="intervalMatrixIndexing"
      at Object.createCompilerError (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:19:19)
      at createDOMCompilerError (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-dom/dist/compiler-dom.cjs.js:2483:25)
      at checkDuplicatedValue (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-dom/dist/compiler-dom.cjs.js:2547:29)
      at transformModel (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-dom/dist/compiler-dom.cjs.js:2579:29)
      at buildProps (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:4777:48)
      at Array.postTransformElement (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:4374:38)
      at traverseNode (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:2225:19)
      at traverseChildren (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:2167:9)
      at traverseNode (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:2218:13)
      at traverseChildren (/home/inthar/Downloads/scale-workshop-main/node_modules/@vue/compiler-core/dist/compiler-core.cjs.js:2167:9)

Comment (inthar-raven) [reply src/App.vue:123]: It's one of these things that seem like they ought to work but don't.

Comment (frostburn) [reply src/App.vue:123]: Yeah, you need an auxillary ref and a watcher if you want to use `v-model` it's better to just emit the correct values from the radio buttons directly.

Comment (inthar-raven) [reply src/App.vue:123]: What can I use to label the data instead of a `v-model`?

File: src/views/AnalysisView.vue [hunk 9]
```diff
@@ -143,6 +154,28 @@ const matrix = computed(() => {
           <label for="format-decimal"> Decimal ratio </label>
         </span>
       </div>
+      <div class="control radio-group">
+        <label>Interval indexing</label>
+        <span>
+          <input
+            type="number"
+            id="indexing-zero"
+            value="0"
```
Comment (frostburn) [inline src/views/AnalysisView.vue:163]: Sorry about before. Your problem is that this should be `:value="0"` instead.

File: src/views/AnalysisView.vue [hunk 10]
```diff
@@ -143,6 +154,28 @@ const matrix = computed(() => {
...
+            v-model="intervalMatrixIndexing"
+          />
+          <label for="indexing-zero"> 0-indexing (default) </label>
+        </span>
+        
+        <span>
+          <input
+            type="number"
+            id="indexing-one"
+            value="1"
```
Comment (frostburn) [inline src/views/AnalysisView.vue:173]: And this `:value="1"` then the radio buttons work with numbers just fine.

File: src/views/AnalysisView.vue [hunk 11]
```diff
@@ -87,7 +97,8 @@ const matrix = computed(() => {
     props.scale.head(MAX_SCALE_SIZE).mergeOptions({
       centsFractionDigits: 1,
       decimalFractionDigits: 3,
-    })
+    }), 
+    0
```
Comment (frostburn) [inline src/views/AnalysisView.vue:101]: This whole change should be reverted as I misread your intention when suggesting it.

**Comment by frostburn** [comment]
This is now 13 commits long. They should all be squashed into one so that we get a readable commit history in `main`.

File: src/analysis.ts [hunk 12]
```diff
@@ -105,10 +105,10 @@ export function utonalFundamental(frequencies: number[], maxDivisor = 23) {
 // Interval matrix a.k.a the modes of a scale
 export function intervalMatrix(scale: Scale) {
   const result = [];
-  const degrees = [...Array(scale.size + 1).keys()];
+  const columns = [...Array(scale.size + 1).keys()];
```
Comment (frostburn) [inline src/analysis.ts:108]: A little unrelated, but OK...

File: src/App.vue [hunk 13]
```diff
@@ -783,7 +784,9 @@ onMounted(() => {
   if ("midiOctaveOffset" in storage) {
     midiOctaveOffset.value = parseInt(storage.getItem("midiOctaveOffset")!);
   }
-
+  if ("intervalMatrixIndexing" in storage) {
+    intervalMatrixIndexing.value = parseInt(storage.getItem("intervalMatrixIndexing")!);
```
Comment (frostburn) [inline src/App.vue:788]: `parseInt(null)` results in `NaN` so that exclamation mark is definitely in the wrong here. (Remember to test after running `window.localStorage.clear()` in the console.)

**Comment by frostburn** [comment]
GitHubActions seems to have something to say about your code style.

**Comment by frostburn** [comment]
The commit message is too long. Use `git commit --amend` to reword it. (Also handy to avoid generating new commits so you avoid the rebases).

File: src/App.vue [hunk 14]
```diff
@@ -783,7 +784,11 @@ onMounted(() => {
   if ("midiOctaveOffset" in storage) {
     midiOctaveOffset.value = parseInt(storage.getItem("midiOctaveOffset")!);
   }
-
+  if ("intervalMatrixIndexing" in storage) {
+    intervalMatrixIndexing.value = parseInt(
+      storage.getItem("intervalMatrixIndexing")!
```
Comment (frostburn) [inline src/App.vue:789]: Still doing naughty things and storing a `NaN` on first load.

**Comment by frostburn** [comment]
Commits need squashing again.

File: src/App.vue (hunk 14 above)
Comment (frostburn) [reply src/App.vue:789]: This really should be `parseInt(storage.getItem("intervalMatrixIndexing") ?? "0", 10)`

**Comment by frostburn** [comment]
Feel free to add your name on the About page.
